vmap: temporarily disable support for (most) random functions

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#42324 vmap: temporarily disable support for (most) random functions**
* #42248 Implement batching rules for some view ops
* #41943 vmap fallback kernel

While we figure out the random plan, I want to initially disable
support for random operations. This is because there is an ambiguity in
what randomness means. For example,
```
tensor = torch.zeros(B0, 1)
vmap(lambda t: t.normal_())(tensor)
```
in the above example, should `tensor[0]` and `tensor[1]` be equal (i.e.,
use the same random seed), or should they be different?

The mechanism for disabling random support is as follows:
- we register a batching rule that throws an error
- for any future (out-of-place) random operations, the fallback path
will probably catch them and throw an error.

I want to disable random factory functions as well, but I haven't
figured out how to do that yet. (e.g., torch.rand, torch.randn). That
will come next.

Test Plan:
- `pytest test/test_vmap.py -v -k "Operators"`

Differential Revision: [D22846274](https://our.internmc.facebook.com/intern/diff/D22846274)

zou3519 committed:
vmap: temporarily disable support for (most) random functions

While we figure out the random plan, I want to initially disable
support for random operations. This is because there is an ambiguity in
what randomness means. For example,
```
tensor = torch.zeros(B0, 1)
vmap(lambda t: t.normal_())(tensor)
```
in the above example, should `tensor[0]` and `tensor[1]` be equal (i.e.,
use the same random seed), or should they be different?

The mechanism for disabling random support is as follows:
- we register a batching rule that throws an error
- for any future (out-of-place) random operations, the fallback path
will probably catch them and throw an error.

I want to disable random factory functions as well, but I haven't
figured out how to do that yet. (e.g., torch.rand, torch.randn). That
will come next.

Test Plan:
- `pytest test/test_vmap.py -v -k "Operators"`

[ghstack-poisoned]

ezyang commented on aten/src/ATen/BatchedFallback.cpp:
```diff
@@ -33,6 +33,22 @@ static bool areAnyArgumentsTensorList(const FunctionSchema& schema) {
+      schema.arguments().begin(),
+      schema.arguments().end(),
+      [] (const Argument& arg) {
+        // Check for Generator or Optional<Generator>.
```
If someone ever adds list of generators I will eat my shoe.

ezyang commented on aten/src/ATen/BatchingRegistrations.cpp:
```diff
@@ -165,6 +165,17 @@ Tensor diagonal_batching_rule(const Tensor& self, int64_t offset, int64_t dim1,
+template <typename... Args> Tensor& unsupportedRandomOp_(Args... args) {
+  TORCH_CHECK(false, "vmap: We do not yet support random operations. Please ",
+              "perform random operations outside of vmap as a workaround");
+}
```
You could do this with marginally less boilerplate by making another magical unsupported boxed kernel, similar to the named tensor not supported kernel. Then it wouldn't be necessary to get the correct type instantiations below. Up to you.

zou3519 replied:
I'll add that as a follow-up item to investigate. Getting the correct type instantiations is annoying (and potentially fragile if someone updates the schema) so the magical unsupported kernel sounds nice.

ezyang commented on test/test_vmap.py:
```diff
@@ -596,6 +596,37 @@ def test_t(self):
+        B0 = 2
+        random_ops = [
+            # out-of-place
+            (torch.bernoulli, (torch.rand(B0, 1),)),
```
@mruberry Op DB!!!

zou3519 replied:
an op database would indeed solve a lot of my testing problems

zou3519 committed:
Update on "vmap: temporarily disable support for (most) random functions"


While we figure out the random plan, I want to initially disable
support for random operations. This is because there is an ambiguity in
what randomness means. For example,
```
tensor = torch.zeros(B0, 1)
vmap(lambda t: t.normal_())(tensor)
```
in the above example, should `tensor[0]` and `tensor[1]` be equal (i.e.,
use the same random seed), or should they be different?

The mechanism for disabling random support is as follows:
- we register a batching rule that throws an error
- for any future (out-of-place) random operations, the fallback path
will probably catch them and throw an error.

I want to disable random factory functions as well, but I haven't
figured out how to do that yet. (e.g., torch.rand, torch.randn). That
will come next.

Test Plan:
- `pytest test/test_vmap.py -v -k "Operators"`

Differential Revision: [D22846274](https://our.internmc.facebook.com/intern/diff/D22846274)

[ghstack-poisoned]

zou3519 committed:
Update on "vmap: temporarily disable support for (most) random functions"


While we figure out the random plan, I want to initially disable
support for random operations. This is because there is an ambiguity in
what randomness means. For example,
```
tensor = torch.zeros(B0, 1)
vmap(lambda t: t.normal_())(tensor)
```
in the above example, should `tensor[0]` and `tensor[1]` be equal (i.e.,
use the same random seed), or should they be different?

The mechanism for disabling random support is as follows:
- we register a batching rule that throws an error
- for any future (out-of-place) random operations, the fallback path
will probably catch them and throw an error.

I want to disable random factory functions as well, but I haven't
figured out how to do that yet. (e.g., torch.rand, torch.randn). That
will come next.

Test Plan:
- `pytest test/test_vmap.py -v -k "Operators"`

Differential Revision: [D22846274](https://our.internmc.facebook.com/intern/diff/D22846274)

[ghstack-poisoned]

zou3519 committed:
Update on "vmap: temporarily disable support for (most) random functions"


While we figure out the random plan, I want to initially disable
support for random operations. This is because there is an ambiguity in
what randomness means. For example,
```
tensor = torch.zeros(B0, 1)
vmap(lambda t: t.normal_())(tensor)
```
in the above example, should `tensor[0]` and `tensor[1]` be equal (i.e.,
use the same random seed), or should they be different?

The mechanism for disabling random support is as follows:
- we register a batching rule that throws an error
- for any future (out-of-place) random operations, the fallback path
will probably catch them and throw an error.

I want to disable random factory functions as well, but I haven't
figured out how to do that yet. (e.g., torch.rand, torch.randn). That
will come next.

Test Plan:
- `pytest test/test_vmap.py -v -k "Operators"`

Differential Revision: [D22846274](https://our.internmc.facebook.com/intern/diff/D22846274)

[ghstack-poisoned]

zou3519 commented:
I'm going to hold off on landing this until I figure out https://github.com/pytorch/pytorch/pull/42324#discussion_r463299167. There's a case where this PR doesn't cover and a dispatcher-approach would (we also want to ban random operations for regular tensors (non-BatchedTensors) inside vmap!)

zou3519 commented:
#42617 supercedes this.
